controller/token.go

```diff
@@ -353,115 +354,200 @@ func (c *TokenController) Exchange(ctx *app.ExchangeTokenContext) error {
-			if bcrypt.CompareHashAndPassword([]byte(hash), secret) == nil {
-				tokenType := "bearer"
-				accessToken, err := c.TokenManager.GenerateServiceAccountToken(ctx.RequestData, sa.ID, sa.Name)
-				if err != nil {
-					return jsonapi.JSONErrorResponse(ctx, err)
-				}
-				pat := &app.OauthToken{
-					AccessToken: &accessToken,
-					TokenType:   &tokenType,
-				}
-				return ctx.OK(pat)
-			}
-		}
+	return ctx.OK(token)
+}
+
+func (c *TokenController) exchangeWithGrantTypeRefreshToken(ctx *app.ExchangeTokenContext) (*app.OauthToken, error) {
+
+	payload := ctx.Payload
+	refreshToken := payload.RefreshToken
+	if refreshToken == nil {
+		return nil, errors.NewBadParameterError("refresh_token", nil).Expected("not nil")
+	}
+
+	// Default value of this public client id is set to "740650a2-9c44-4db5-b067-a3d1b2cd2d01"
+	if payload.ClientID != c.Configuration.GetPublicOauthClientID() {
 		log.Error(ctx, map[string]interface{}{
-			"client_id":     payload.ClientID,
-			"client_secret": *payload.ClientSecret,
-		}, "Service Account secret doesn't match")
-		return jsonapi.JSONErrorResponse(ctx, errors.NewUnauthorizedError("invalid Service Account ID or secret"))
-	} else if payload.GrantType == authorizationCode {
-		if payload.Code == nil {
-			return jsonapi.JSONErrorResponse(ctx, errors.NewBadParameterError("code", "nil").Expected("authorization code"))
-		}
-		// Default value of this public client id is set to "740650a2-9c44-4db5-b067-a3d1b2cd2d01"
-		if payload.ClientID != c.Configuration.GetPublicOauthClientID() {
-			log.Error(ctx, map[string]interface{}{
-				"client_id": payload.ClientID,
-			}, "unknown oauth client id")
-			return jsonapi.JSONErrorResponse(ctx, errors.NewUnauthorizedError("invalid oauth client id"))
-		}
-		authEndpoint, err := c.Configuration.GetKeycloakEndpointAuth(ctx.RequestData)
-		if err != nil {
-			log.Error(ctx, map[string]interface{}{
-				"err": err,
-			}, "unable to get keycloak auth endpoint url")
-			return jsonapi.JSONErrorResponse(ctx, errors.NewInternalError(ctx, errs.Wrap(err, "unable to get keycloak auth endpoint url")))
-		}
+			"client_id": payload.ClientID,
+		}, "unknown oauth client id")
+		return nil, errors.NewUnauthorizedError("invalid oauth client id")
+	}
 
-		tokenEndpoint, err := c.Configuration.GetKeycloakEndpointToken(ctx.RequestData)
-		if err != nil {
-			log.Error(ctx, map[string]interface{}{
-				"err": err,
-			}, "unable to get keycloak token endpoint url")
-			return jsonapi.JSONErrorResponse(ctx, errors.NewInternalError(ctx, errs.Wrap(err, "unable to get keycloak token endpoint url")))
-		}
+	client := &http.Client{Timeout: 10 * time.Second}
+	endpoint, err := c.Configuration.GetKeycloakEndpointToken(ctx.RequestData)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"err": err,
+		}, "Unable to get Keycloak token endpoint URL")
+		return nil, errors.NewInternalError(ctx, errs.Wrap(err, "unable to get Keycloak token endpoint URL"))
```

Comment:
there's a method called `NewInternalErrorWithString..()` which could be used here.

Reply:
https://github.com/fabric8-services/fabric8-auth/pull/272/commits/5e6dd11400c637205d052e67775aee8e3299e378